lib/remote.go

```diff
@@ -76,58 +83,105 @@ func (r *RemoteRequests) PushToRemote(p *PushParams, out *bool) error {
 		return err
 	}
 
-	content, err := ioutil.ReadAll(res.Body)
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("error code %d: %v", res.StatusCode, rejectionReason(res.Body))
+	}
+
+	// TODO(dlong): Get a dsync session id from the remote, use it to perform dsync.
+
+	ctx := context.Background()
+
+	ng, err := newNodeGetter(r.node)
 	if err != nil {
-		panic(err)
+		return err
 	}
-	// TODO(dlong): Inspect the remote's response, and then perform dsync.
-	fmt.Printf(string(content))
 
-	if res.StatusCode != http.StatusOK {
-		return fmt.Errorf("error code %d", res.StatusCode)
+	remote := &dsync.HTTPRemote{
+		URL: fmt.Sprintf("%s/dsync", location),
 	}
 
+	send, err := dsync.NewSend(ctx, ng, dinfo.Manifest, remote)
+	if err != nil {
+		return err
+	}
+
+	err = send.Do()
+	if err != nil {
+		return err
+	}
+
+	// TODO(dlong): Pin the data.
+
 	*out = true
 	return nil
 }
 
+// newNodeGetter generates an ipld.NodeGetter from a QriNode
+func newNodeGetter(node *p2p.QriNode) (ng ipld.NodeGetter, err error) {
+	ipfsn, err := node.IPFSNode()
+	if err != nil {
+		return nil, err
+	}
+
+	ng = &dag.NodeGetter{Dag: coreapi.NewCoreAPI(ipfsn).Dag()}
+	return
+}
+
 // Receive is used to save a dataset when running as a remote. API only, not RPC or command-line.
-func (r *RemoteRequests) Receive(p *ReceiveParams, out *bool) (err error) {
+func (r *RemoteRequests) Receive(p *ReceiveParams, reason *string) (err error) {
 	if r.cli != nil {
 		return fmt.Errorf("receive cannot be called over RPC")
 	}
 
-	dinfo := dagInfoTmp{}
+	dinfo := dag.Info{}
 	err = json.Unmarshal([]byte(p.Body), &dinfo)
 	if err != nil {
 		return err
 	}
 
-	fmt.Printf("Received dag.Info:\n")
-	fmt.Printf(p.Body)
-	fmt.Printf("\n\n")
+	// TODO(dlong): Customization for how to decide to accept the dataset.
+	if !Config.API.RemoteAlwaysAccept {
+		*reason = "not accepting any datasets"
```

Comment:
If you're ok with the max size adjustment I think this message should stay the exact same if `remoteMaxSizeAccept == 0`